### PR TITLE
Add Iterator::find_map

### PR DESCRIPTION
**Pull request by matklad**
I'd like to propose to add `find_map` method to the `Iterator`: an occasionally useful utility, which relates to `filter_map` in the same way that `find` relates to `filter`.

`find_map` takes an `Option`-returning function, applies it to the elements of the iterator, and returns the first non-`None` result. In other words, `find_map(f) == filter_map(f).next()`.

Why do we want to add a function to the `Iterator`, which can be trivially expressed as a combination of existing ones? Observe that `find(f) == filter(f).next()`, so, by the same logic, `find` itself is unnecessary!

The more positive argument is that desugaring of  `find[_map]` in terms of `filter[_map]().next()` is not super obvious, because the `filter` operation reads as if it is applies to the whole collection, although in reality we are interested only in the first element. That is, the jump from "I need a **single** result" to "let's use a function which maps **many** values to **many** values" is a non-trivial speed-bump, and causes friction when reading and writing code.

Does the need for `find_map` arise in practice? Yes! 

* Anecdotally, I've more than once searched the docs for the function with `[T] -> (T -> Option<U>) -> Option<U>` signature. 
* The direct cause for this PR was [this](https://github.com/rust-lang/cargo/pull/5187/files/1291c50e86ed4b31db0c76de03a47a5d0074bbd7#r174934173) discussion in Cargo, which boils down to "there's some pattern that we try to express here, but current approaches looks non-pretty" (and the pattern is `filter_map`
* There are several `filter_map().next` combos in Cargo: [[1]](https://github.com/rust-lang/cargo/blob/545a4a2c930916cc9c3dc1716fb7a33299e4062b/src/cargo/ops/cargo_new.rs#L585), [[2]](https://github.com/rust-lang/cargo/blob/545a4a2c930916cc9c3dc1716fb7a33299e4062b/src/cargo/core/resolver/mod.rs#L1130), [[3]](https://github.com/rust-lang/cargo/blob/545a4a2c930916cc9c3dc1716fb7a33299e4062b/src/cargo/ops/cargo_rustc/mod.rs#L1086).
* I've also needed similar functionality in `Kotlin` several times. There, it is expressed as `mapNotNull {}.firstOrNull`, as can be seen [here](https://github.com/intellij-rust/intellij-rust/blob/ee8bdb4e073fd07142fc6e1853ca288c57495e69/src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt#L154), [here](https://github.com/intellij-rust/intellij-rust/blob/ee8bdb4e073fd07142fc6e1853ca288c57495e69/src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt#L444) [here](https://github.com/intellij-rust/intellij-rust/blob/ee8bdb4e073fd07142fc6e1853ca288c57495e69/src/main/kotlin/org/rust/ide/inspections/RsLint.kt#L38) and [here](https://github.com/intellij-rust/intellij-rust/blob/ee8bdb4e073fd07142fc6e1853ca288c57495e69/src/main/kotlin/org/rust/cargo/toolchain/RustToolchain.kt#L74) (and maybe in some other cases as well)

Note that it is definitely not among the most popular functions (it definitely is less popular than `find`), but, for example it (in case of Cargo) seems to be more popular than `rposition` (1 occurrence), `step_by` (zero occurrences) and `nth` (three occurrences as `nth(0)` which probably should be replaced with `next`).

Do we necessary need this function in `std`? Could we move it to itertools? That is possible, but observe that `filter`, `filter_map`, `find` and `find_map` together really form a complete table:

|||
|-------|---------|
| filter| find|
|filter_map|find_map|

It would be somewhat unsatisfying to have one quarter of this table live elsewhere :) Also, if `Itertools` adds an `find_map` method, it would be more difficult to move it to std due to name collision.

Hm, at this point I've searched for `filter_map` the umpteenth time, and, strangely, this time I do find this RFC: https://github.com/rust-lang/rfcs/issues/1801. I guess this could be an implementation though? :)

To sum up:

Pro:
  - complete the symmetry with existing method
  - codify a somewhat common non-obvious pattern 

Contra:
  - niche use case
  - we can, and do, live without it

### CODE REVIEW

**Comment by Centril** [comment]
I think I'm for this PR, but a small note re:

> Also, if `Itertools` adds an `find_map` method, it would be more difficult to move it to std due to name collision.

I think given the latest `.flatten()` debacle, we kinda informally "decided" that `Iterator` extension proposals should first be proposed to libstd, and if libstd doesn't want it, then it should be proposed for `Itertools` inclusion.

**Comment by leonardo-m** [comment]
Instead of adding iterators randomly, I suggest to take a look at the most commonly used iterators from the iterator crate, and add those.

**Comment by Centril** [comment]
> Instead of adding iterators randomly,

There's nothing random about this addition - I think it is right to fill consistency holes per @matklad's table above.

That said, we should add some operations from `Itertools` - any suggestions?

**Comment by leonardo-m** [comment]
By "randomly" I meant: "not needed in lot of code by lot of people". I agree it's hard to measure this need exactly, but even measuring it approximately is OK. I have written a ton of iterator-based Rust/Python/D code and I think there are things (much) more useful than a find_map.

Regarding the list of "most useful itertools things", I can give you a list, but probably every programmer has a slightly different list. In my opinion the intersection of such lists is far from empty. I think this topic is better discussed in a forum thread than here.

**Comment by Centril** [comment]
> Regarding the "most useful itertools things", I can give you a list,

Please do =) (and also discount things that require allocation, because those can't be in libcore)

**Comment by matklad** [comment]
@leonardo-m I totally agree that “how often is this used in practice” should be one of the primarily axis along which to measure additions to std. That’s why I’ve tried to measure how often could find_map be used in practice, by giving examples of real-life code :)

>I have written a ton of iterator-based Rust/Python/D code and I think there are things (much) more useful than a find_map.

Keep in mind that what kinds of iterators you need most highly depends on the kind of code you are writing. For me, find_map is the *second* most important missing method at the moment (the first one is join of course, producing a comma separated list of things in Rust is ridiculously not straight forward at the moment)

**Comment by shepmaster** [comment]
Ping from triage, @KodrAus ! Will you have time to review this soon?

**Comment by KodrAus** [comment]
This particular `find_map` addition looks reasonable to me, since it rounds out some existing set of methods as @matklad described.

cc @rust-lang/libs any thoughts on `find_map`? If we're generally happy with landing this then maybe we can throw a test case or two in there.

**Comment by matklad** [comment]
Pushed some tests!

**Review by KodrAus — APPROVED** [review]
Thanks @matklad! I think this has been sitting around long enough. Once we have the unstable issue number sorted I think this will be ready to go.

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -1745,6 +1745,38 @@ pub trait Iterator {
...
+    /// let a = ["lol", "NaN", "2", "5"];
+    ///
+    /// let mut first_number = a.iter().find_map(|s| s.parse().ok());
+    ///
+    /// assert_eq!(first_number, Some(2));
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_find_map",
+               reason = "unstable new API",
+               issue = "0")]
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/libcore/iter/iterator.rs]: Filled in the issue number!

File: src/libcore/iter/iterator.rs [hunk 2]
```diff
@@ -1745,6 +1745,38 @@ pub trait Iterator {
...
+    /// `iter.find_map(f)` is equivalent to `iter.filter_map(f).next()`.
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iterator_find_map)]
+    /// let a = ["lol", "NaN", "2", "5"];
+    ///
+    /// let mut first_number = a.iter().find_map(|s| s.parse().ok());
```
Comment (meven) [inline src/libcore/iter/iterator.rs]: The `mut` is not needed here, I believe.

Comment (shepmaster) [reply src/libcore/iter/iterator.rs]: @meven seems plausible — submit a PR to fix it, please?

Comment (meven) [reply src/libcore/iter/iterator.rs]: PR opened at https://github.com/rust-lang/rust/pull/55389
